### DISCUSSION TITLE
Using createListenerMiddleware per slice

### DISCUSSION THREAD

**Discussion by jd-carroll** [reduxjs/redux-toolkit#3665] [General]
We have multiple slices (> 10 but < 20) and we would love to use `createListenerMiddleware` within each slice.

I'm assuming it is safe to call `createListenerMiddleware()` in each slice and then have the store do something like:
```jsx
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().prepend(...arrayWithAllMiddlewareListeners),
```

Again, I'm assuming this is safe, but thought I would check just to be sure.

**Comment by markerikson** [comment]
No, don't do that!  You should have _one_ instance of the listener middleware in the app's store.  Having multiple copies is both redundant, _and_ adding a lot of extra overhead for every dispatched action.

Just add the one listener middleware instance, then add as many listener entries as you need throughout the app:

- https://redux-toolkit.js.org/api/createListenerMiddleware#organizing-listeners-in-files

(Just out of curiosity, what sort of "per-slice"-type behavior are you wanting to add?)

**Comment by jd-carroll** [comment]
Thank you, that was actually my specific concern.

It's probable that we are abusing slices... We are creating a separate slice for each defined feature area of the app. A handful of the slices will have 10+ (but < 30) reducers/actions, while all the others are definitely <10 reducers / actions. Again, its mainly to narrow the focus of each "feature" than anything else.

**Comment by jd-carroll** [comment]
And I should have finished reading the docs... Thanks for the link

Curious to hear if you think I'm being ridiculous with my slice configuration, otherwise I'll mark this as closed

**Reply by markerikson** [reply]
We do recommend "slice-per-feature" as the general approach, yeah:

- https://redux.js.org/style-guide/#structure-files-as-feature-folders-with-single-file-logic

what "feature" means specifically is up to you and your app, but that sounds like it matches what we're suggesting.
